### PR TITLE
[firefox-android: jhugman/bug-1841262-add-coenrolling-features-to-nimbus-builder] Add API to SDK populating EnrollmentsEvolver coenrolling feature ids

### PR DESCRIPTION
**Pull request by jhugman**
Fixes [EXP-3623](https://mozilla-hub.atlassian.net/browse/EXP-3623).

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [x] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/NimbusBuilder.kt [hunk 1]
```diff
@@ -59,6 +58,26 @@ abstract class AbstractNimbusBuilder<T : NimbusInterface>(val context: Context)
...
+    /**
+     * Optional callback to be called everytime experiments have been fetched.
+     *
+     * Experiment recipes are usually fetched shortly after startup, and those pending recipes are
+     * applied at the following startup.
+     *
+     * This is not usually needed.
+     */
+    var onFetchCallback: () -> Unit = {}
+
```
Comment (jhugman) [inline components/nimbus/android/src/main/java/org/mozilla/experiments/nimbus/NimbusBuilder.kt:80]: Knowing this was a breaking change, I took the time to refactor the NimbusBuilder for Android and pull in the fetchCallback and applyCallback so they work in the same for when in Android Components and just with the DefaultNimbusBuilder in Application Services.

File: components/nimbus/ios/Nimbus/NimbusBuilder.swift [hunk 2]
```diff
@@ -177,6 +190,15 @@ public class NimbusBuilder {
...
             }
 
+            if let callback = onFetchCallback {
+                NotificationCenter.default.addObserver(forName: .nimbusExperimentsFetched,
+                                                       object: nil,
+                                                       queue: nil)
+                { _ in
+                    callback(nimbus)
+                }
+            }
```
Comment (jeddai) [inline components/nimbus/ios/Nimbus/NimbusBuilder.swift:200]: I'm not familiar with the notification center piece here, can you elaborate on what that's doing?

**Review by jeddai — APPROVED** [review]
r+wc

One comment with a change request and one question

File: components/nimbus/ios/Nimbus/NimbusBuilder.swift [hunk 3]
```diff
@@ -108,6 +108,18 @@ public class NimbusBuilder {
 
     var onApplyCallback: ((NimbusInterface) -> Void)?
 
+    /**
+     * Optional callback to be called after the fetch of new experiments has completed.
+     * experiments recipes.
+     */
+    @discardableResult
+    public func onFetch(callback: @escaping (NimbusInterface) -> Void) -> NimbusBuilder {
+        onApplyCallback = callback
```
[the remark this answers is not in the document]
Comment (jhugman) [reply components/nimbus/ios/Nimbus/NimbusBuilder.swift:117]: Oh yes! Great catch!

**Comment by tarikeshaq** [comment]
@jhugman This landed before the branch builds finished since they don't block the merge - looks like the branch builds failed here - just a heads up for when the iOS/Fenix bumps are up tomorrow

**Pull request merged**